File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
     Map<Character,String>map;
     List<String>res;
    public List<String> letterCombinations(String digits) {
       map=new HashMap<>();
        res=new ArrayList<>();
      
        map.put('2',"abc");
        map.put('3',"def");
        map.put('4',"ghi");
        map.put('5',"jkl");
        map.put('6',"mno");
        map.put('7',"pqrs");
        map.put('8',"tuv");
        map.put('9',"wxyz");
                StringBuilder sb=new StringBuilder();
        
        back(digits.toCharArray(),0,sb);
        
        return res;
        
    }
   void back(char[]ch,int pos,StringBuilder sb){
       if(pos>=ch.length)return;
       
       for(char c:map.get(ch[pos]).toCharArray()){
            sb.append(c);
           back(ch,pos+1,sb);
          if(sb.length()==ch.length)
           res.add(sb.toString());
           sb.deleteCharAt(sb.length()-1);
       }
       
   }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp

class Solution {
private:
    void helper(vector<string>&ans, unordered_map<char,string>&btns, string &digits,string &tmp, int idx){
        int n = digits.length();
        
        if(tmp.length()==n){
            ans.push_back(tmp);
        }
        if(idx>=n){
            return ;
        }
         for(auto c : btns[digits[idx]]){
             tmp += c;
             helper(ans,btns,digits,tmp,idx+1);
             tmp.pop_back();
        }
    }
public:
    vector<string> letterCombinations(string digits) {
        if(digits.size()==0){
            return vector<string>{};
        }
        unordered_map<char,string>btns;
        btns['2'] = "abc";
        btns['3'] = "def";
        btns['4'] = "ghi";
        btns['5'] = "jkl";
        btns['6'] = "mno";
        btns['7'] = "pqrs";
        btns['8'] = "tuv";
        btns['9'] = "wxyz";
        vector<string> ans;
        string tmp = "";
        helper(ans, btns, digits, tmp, 0);
       // cout<<"ans.szie "<<ans.size()<<"\n";
        return  ans;
    }
};
